src/hopen/core.cljc

```diff
@@ -19,25 +19,70 @@
             :else element))]
    (f-eval element)))
 
+;; Top level evaluation functions
+;; The given element can be a "block" to be executed or an simple expression to be evaluated
 (defn- rf-block [rf env result element]
   (or (when (list? element)
         (let [[f-symb & args] element
-              f (get-in env [:block-macro f-symb])]
-          (when f
+              f (get-in env [:block-macro f-symb])
+
+              ;; Values in :block-macro could be a function or a var
+              ;; If it's a var, resolve it now (into a function).
+              ;; This allows us to frequently rebind/re-eval functions without
+              ;; having to reconstruct `default-env`
+              f (if-not (var? f)
+                  f
+                  (var-get f))]
+          (when (fn? f)
             (apply f rf env result args))))
       (rf result (tpl-eval env element))))
 
 (defn- rf-for [rf env result bindings content]
   (if (seq bindings)
-    (let [[k coll & next-bindings] bindings]
-      (reduce (fn [result val]
-                (rf-for rf
-                        (update env :bindings assoc k val)
-                        result
-                        next-bindings
-                        content))
-              result
-              (tpl-eval env coll)))
+    ;; Deconstruct one pair of the `for` binding
+    (let [[k coll & next-bindings] bindings
```

Comment:
I was hoping this code for parsing binding sequences to be in a separate function.

Also, it would be nice to have it work in a general way, so that it would still work the day we choose to add more qualifiers on bindings for the `for` or `let` functions.

I am going to add a commit to specify how that function would work via some tests, please wait a few minutes.

Reply:
Okay... it'd be pretty straightforward to pull out this and the next few lines that deals with qualifiers/clauses into something more generic.

Maybe something along the lines of...
```
(defn extract-single-binding [bindings]) -> [symb coll {:qualifier-name qualifier-val ...} next-bindings]
```

Reply:
Implemented in 4829d3c